Remove Enums in favor of TS types with guard functions

ericdnielsen opened the pull request:
Remove our Enums in favor of the TS types (string unions), and guard functions that enforce those values.  The TS types still grant internal library strong typing, while the guard functions work both in JS and TS.  The elimination of string<->Enum conversions during serialization allows removal of some repeated duplication of string constants an simplification of serialization infrastructure.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-layers/164)
<!-- Reviewable:end -->

caisq commented:
Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---

*[src/layers/wrappers.ts, line 240 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-LBw9lip2978JaDEf5ys:-LBw9lip2978JaDEf5yt:b-ij451v) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/897882f8354dd5351d6f420f07f891cf7b8570df/src/layers/wrappers.ts#L240)):*
> ```
> undefined, null
> ```

Are these two entries necessary, given your `== null` check below?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164)*
<!-- Sent from Reviewable.io -->

ericdnielsen commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/layers/wrappers.ts, line 240 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-LBw9lip2978JaDEf5ys:-LBwBNIl49zvr8hM4nW9:b-l1m41f) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/897882f8354dd5351d6f420f07f891cf7b8570df/src/layers/wrappers.ts#L240)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> undefined, null
> ```

Are these two entries necessary, given your `== null` check below?
</blockquote></details>

For all the other enum-like things we have undefined and null in the VALID_* list.  Partially because the error message wants to show that undefined/null are allowed values.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164)*
<!-- Sent from Reviewable.io -->

caisq commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/layers/wrappers.ts, line 243 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-LBwCRnm-o_95M_5kM42:-LBwCRnm-o_95M_5kM43:b-b69go0) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/897882f8354dd5351d6f420f07f891cf7b8570df/src/layers/wrappers.ts#L243)):*
> ```
>   if (value == null) {
>     return;
>   }
> ```

Given what you said above, this check here is redundant, I think.

---

*[src/layers/wrappers.ts, line 247 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-LBwCYyi-o8yEQeQEF3K:-LBwCYyi-o8yEQeQEF3L:bhfegcj) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/897882f8354dd5351d6f420f07f891cf7b8570df/src/layers/wrappers.ts#L247)):*
> ```
> as
> ```

s/as/are/

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-:-LBwCf7c3oJBgxk-BAVW:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

ericdnielsen commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[src/layers/wrappers.ts, line 243 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-LBwCRnm-o_95M_5kM42:-LBwDlsS8ORD_cg80EdJ:b53omhv) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/897882f8354dd5351d6f420f07f891cf7b8570df/src/layers/wrappers.ts#L243)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
>   if (value == null) {
>     return;
>   }
> ```

Given what you said above, this check here is redundant, I think.
</blockquote></details>

Removed here, and on the other check* functions.

---

*[src/layers/wrappers.ts, line 247 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-LBwCYyi-o8yEQeQEF3K:-LBwDisY7MGqGWIib2ai:bb9n5t8) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/897882f8354dd5351d6f420f07f891cf7b8570df/src/layers/wrappers.ts#L247)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> as
> ```

s/as/are/
</blockquote></details>

Done, and corrected on the other check* functions.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164)*
<!-- Sent from Reviewable.io -->

davidsoergel commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 4 of 7 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-:-LBwJ8x_0CNBZ9UnbPwZ:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

ericdnielsen commented:
Shanqing and I had some discussions post LGTM that led to a rather significant refactor.  PR is updated, I'll hold off on submitting until the morning at least.  Just in case there's comments on the updated version.

---

Review status: 2 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-:-LBwYGzsAKvcEXPNqmg-:blenibw)*
<!-- Sent from Reviewable.io -->

caisq commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 2 of 9 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/164#-:-LBzw6Wa3UsSLK6N0hq1:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
